React/forms.md

```diff
@@ -2,24 +2,280 @@
+      name: event.target.value,
+    });
+  }
+```
+
+Then add `onChange` and `value` fields to the `input` in `render`.
+
+```javascript
+<input
+  onChange={ this.onNameChange }
+  value={ this.state.name }
+  name="name"
+/>
+```
+
+Now every time the user types into the name input field the `NewStudentForm`'s state is updated.
+
+**Practice**:  Update the e-mail field to have the `NewStudentForm` component manage it's state as well.
+
+**Question** Why would it be useful to track the state of a form field?
+
+
+## Form Validation
+
+By allowing the `NewStudentForm` component track the status of the form fields this will allow us to:
+- Validate form fields on the fly
+- More easily access the form fields when the form is submitted.
+
+We can perform a validation on the email field with a function like this:
+
+```javascript
+emailValid = () => {
+  return this.state.email.match(/\S+@\S+/);
+}
+```
+
+And give the user feedback on validation with:
+
+```javascript
+<input
+  onChange={ this.handleEmailChange }
+  value={ this.state.email }
+  className={ this.emailValid() ? "valid": "invalid" }
+  name="email"
+/>
+```
+
+**Question:**  What does this line with `className=` do?
+
+The form is rerendered every time the state of the component changes, and this code will run `this.emailValid()` and if the email field is valid the input will have a class of `valid`, and if not it will have the class of `invalid`.  With a little CSS we can give the user valuable feedback as to the status of a form field.
+
+```css
+.valid {
+  background: lightgreen;
+}
 
+.invalid {
+  background: pink;
+}
+```
 
+So with our form we can track input into the fields and provide real-time validation feedback.  Next we want to actually **do** something with the data, but our `NewStudentForm` component doesn't, and shouldn't need to care what happens with the submitted data.
+
+**Question**:  Looking at our Ada student list app to date, what component would be managing the student data?
+
+## Handling Submissions
+
+Now we want to handle when the user submits the form.  We can add a function as an event handler.
+
+```javascript
+// new_student_form.js
+...
+handleFormSubmit = (event) => {
+  event.preventDefault();
+  const newStudent = {
+    name: this.state.name,
+    email: this.state.email,
+  };
+
+  this.setState({
+    name: '',
+    email: '',
+  });
+
+  // Now we need to do something with the student...
+}
+```
+
+And in `render` we add an event handler for the submit action to the `form`.
+
+```javascript
+// new_student_form.js
+...
+<form className="new-student-form" onSubmit={this.handleFormSubmit}>
+...
+```
+
+So we have a way to detect submit events on the form, but no way to get data to the rest of the application.  To solve this we will modify the `StudentCollection` component to pass a function as a prop to the `NewStudentForm` so it can pass the student data back.
+
+![passing a callback through props](./images/passing-callback.png)
```

Comment:
Can you put as an HTML comment the location where this image was constructed in case it needs editing later?

Reply:
This is a good diagram. Can we make it a bit bigger and make State lowercase?